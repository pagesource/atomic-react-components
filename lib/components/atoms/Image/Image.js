// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';

import { getLabelFromChildren } from '../../../core/utils/elementLocator';
import styles from './Image.style';

type Props = {
  className: string,
  src: string,
  srcset?: string,
  sizes?: string,
  placeholderSrc?: string,
  alt: string,
  elementLocator?: string,
  inheritedStyles?: string,
  loading?: 'auto' | 'lazy' | 'eager',
  ref: () => void,
};

const Image = ({
  className,
  src,
  srcset,
  sizes,
  placeholderSrc,
  alt,
  inheritedStyles,
  ref,
  elementLocator,
  loading,
  ...other
}: Props): Node => (
  <img
    {...other}
    className={className}
    src={src}
    srcSet={srcset || null}
    sizes={sizes || null}
    alt={alt}
    ref={ref}
    loading={loading}
    data-locator={elementLocator || `image-${getLabelFromChildren(alt)}`}
    onError={
      placeholderSrc
        ? (event) => {
            // eslint-disable-next-line no-param-reassign
            event.target.src = placeholderSrc;
          }
        : null
    }
  />
);

Image.defaultProps = {
  srcset: '',
  sizes: '',
  placeholderSrc: '',
  inheritedStyles: '',
  loading: 'auto',
  elementLocator: '',
};

const StyledImage: ComponentType<Props> = styled(Image)`
  ${styles};
`;

export default StyledImage;

export { Image as ImageVanilla };

export type { Props as ImageProps };
