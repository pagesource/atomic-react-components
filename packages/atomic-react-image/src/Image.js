import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { getLabelFromChildren } from '@xt-pagesource/atomic-react-utils';
import styles from './Image.style';

const Image = ({
  className,
  src,
  srcset,
  sizes,
  placeholderSrc,
  alt,
  inheritedStyles,
  ref,
  elementLocator,
  loading,
  ...other
}) => (
  <img
    {...other}
    className={className}
    src={src}
    srcSet={srcset || null}
    sizes={sizes || null}
    alt={alt}
    ref={ref}
    loading={loading}
    data-locator={elementLocator || `image-${getLabelFromChildren(alt)}`}
    onError={
      placeholderSrc
        ? (event) => {
            // eslint-disable-next-line no-param-reassign
            event.target.src = placeholderSrc;
          }
        : null
    }
  />
);

Image.propTypes = {
  className: PropTypes.string,
  src: PropTypes.string.isRequired,
  srcset: PropTypes.string,
  sizes: PropTypes.string,
  placeholderSrc: PropTypes.string,
  alt: PropTypes.string.isRequired,
  elementLocator: PropTypes.string,
  inheritedStyles: PropTypes.string,
  loading: PropTypes.oneOf(['auto', 'lazy', 'eager']),
  ref: PropTypes.func.isRequired,
};

Image.defaultProps = {
  className: '',
  srcset: '',
  sizes: '',
  placeholderSrc: '',
  inheritedStyles: '',
  loading: 'auto',
  elementLocator: '',
};

const StyledImage = styled(Image)`
  ${styles};
`;

export default StyledImage;

export { Image as ImageVanilla };
