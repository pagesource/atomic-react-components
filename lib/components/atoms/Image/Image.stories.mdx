import { Meta, Story, Preview, Props  } from '@storybook/addon-docs/blocks';
import Image, { ImageVanilla } from '.';
import defaultImage, { lazyLoadedImage, customImageLocator } from './Image.mock';

<Meta title="Atoms/Image" component={ImageVanilla} />


### Image Vanilla
<Preview>
  <Story name="Vanilla Image">
 <ImageVanilla {...defaultImage} />
  </Story>
</Preview>

### custom Image Locator
<Preview>
  <Story name="custom Image">
 <Image {...customImageLocator} />
  </Story>
</Preview>

### lazy Loaded Image
<Preview>
  <Story name="lazy Loaded">
 <Image {...lazyLoadedImage} />
  </Story>
</Preview>

## Props ðŸ”§

<Props of={ImageVanilla} />
