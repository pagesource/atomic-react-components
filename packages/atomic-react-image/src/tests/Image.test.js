import React from 'react';
import { shallow, render } from 'enzyme';
import { ImageVanilla } from '../Image';

describe('Image Component', () => {
  const sampleProps = {
    src: 'http://abc.com',
    placeholderSrc: 'http://xyz.com',
    alt: 'ABC',
    elementLocator: 'my-custom-image-locator',
  };

  test('should render correctly and match snapshots', () => {
    const image = shallow(<ImageVanilla {...sampleProps} />);
    expect(image).toMatchSnapshot();
  });

  test('should change image src on error', () => {
    const image = shallow(<ImageVanilla {...sampleProps} />);
    image.simulate('error', { target: image });
    expect(image.src).toBe(sampleProps.placeholderSrc);
  });

  test('should check the data attribute of image', () => {
    const image = render(<ImageVanilla {...sampleProps} />);

    expect(image.attr('data-locator')).toEqual(sampleProps.elementLocator);
  });

  test('should check the default locator of image', () => {
    const props = { alt: 'my img tag' };
    const customLocator = 'image-my-img';
    const image = render(<ImageVanilla {...props} />);

    expect(image.attr('data-locator')).toEqual(customLocator);
  });
});
