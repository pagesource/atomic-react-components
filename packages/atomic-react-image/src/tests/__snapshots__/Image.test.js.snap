// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component should render correctly and match snapshots 1`] = `
<img
  alt="ABC"
  className=""
  data-locator="my-custom-image-locator"
  loading="auto"
  onError={[Function]}
  sizes={null}
  src="http://abc.com"
  srcSet={null}
/>
`;
